File: src-macos/Sources/CRTCore/UpdateChecker.swift
import Foundation

/// Startup update check against the GitHub releases API (spec §6).
/// Silent on any failure — a flaky connection must never interrupt the app.
public struct UpdateChecker: Sendable {

    public init() {}

    /// Returns the latest release tag when it differs from the running
    /// version, otherwise nil. 5-second timeout.
    public func latestVersionIfNewer() async -> String? {
        guard let url = URL(string: "https://api.github.com/repos/connerglover/crt/releases/latest") else {
            return nil
        }

        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 5
        configuration.timeoutIntervalForResource = 5
        let session = URLSession(configuration: configuration)

        var request = URLRequest(url: url)
        request.setValue(CRTVersion.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return nil
            }
            guard
                let object = try? JSONSerialization.jsonObject(with: data),
                let dict = object as? [String: Any],
                let tag = dict["tag_name"] as? String
            else {
                return nil
            }
            if tag != CRTVersion.current {
                return tag
            }
            return nil
        } catch {
            return nil
        }
    }
}

File: src-windows/CRT.Core/Net/UpdateChecker.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace CRT.Core.Net;

/// <summary>
/// Startup update check — ported from <c>src/crt/updater.py</c>. Silent on any
/// failure: this runs on every launch and a flaky connection must never
/// interrupt using the app.
/// </summary>
public sealed class UpdateChecker
{
    private readonly HttpClient _http;

    public UpdateChecker(HttpClient? http = null)
    {
        _http = http ?? new HttpClient();
        _http.Timeout = TimeSpan.FromSeconds(5);
    }

    /// <summary>
    /// Returns the latest release tag when it differs from the running version,
    /// otherwise null (including on any network/parse failure).
    /// </summary>
    public async Task<string?> CheckForUpdatesAsync(CancellationToken ct = default)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, AppVersion.RepoLatestReleaseApi);
            request.Headers.UserAgent.Add(new ProductInfoHeaderValue("crt", AppVersion.Version));
            using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                return null;
            }
            using var document = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false));
            string? latest = document.RootElement.GetProperty("tag_name").GetString();
            if (!string.IsNullOrEmpty(latest) && latest != AppVersion.Version)
            {
                return latest;
            }
        }
        catch (Exception e) when (e is HttpRequestException or TaskCanceledException or JsonException or KeyNotFoundException or InvalidOperationException)
        {
            // Silent by design.
        }
        return null;
    }
}
